src/compiler/checker.ts

```diff
@@ -13011,6 +13282,69 @@ namespace ts {
+                return cached;
+            }
+            links.resolvedSignature = resolvingSignature;
+
+            let callSignature = resolvedStateLessJsxOpeningLikeElement(openingLikeElement, elementType, candidatesOutArray);
+            if (!callSignature || callSignature === unknownSignature) {
+                const callSignatures = elementType && getSignaturesOfType(elementType, SignatureKind.Call);
+                callSignature = callSignatures[callSignatures.length - 1];
+            }
+            links.resolvedSignature = callSignature;
+            // If signature resolution originated in control flow type analysis (for example to compute the
+            // assigned type in a flow assignment) we don't cache the result as it may be based on temporary
+            // types from the control flow analysis.
+            links.resolvedSignature = flowLoopStart === flowLoopCount ? callSignature : cached;
+            return callSignature;
+        }
+
+        /**
+         * Try treating a given opening-like element as stateless function component and try to resolve a signature.
+         * @param openingLikeElement an JsxOpeningLikeElement we want to try resolve its state-less function if possible
+         * @param elementType a type of the opening-like JSX element, a result of resolving tagName in opening-like element.
+         * @param candidatesOutArray an array of signature to be filled in by the function. It is passed by signature help in the language service;
+         *                           the function will fill it up with appropriate candidate signatures
+         * @return a resolved signature if we can find function matching function signature through resolve call or a first signature in the list of functions.
+         *         otherwise return undefined if tag-name of the opening-like element doesn't have call signatures
+         */
+        function resolvedStateLessJsxOpeningLikeElement(openingLikeElement: JsxOpeningLikeElement, elementType: Type, candidatesOutArray: Signature[]): Signature {
+            if (elementType.flags & TypeFlags.Union) {
+                const types = (elementType as UnionType).types;
+                let result: Signature;
+                types.map(type => {
+                    // This is mainly to fill in all the candidates if there is one.
+                    result = result || resolvedStateLessJsxOpeningLikeElement(openingLikeElement, type, candidatesOutArray);
+                });
+
+                return result;
+            }
+
+            const callSignatures = elementType && getSignaturesOfType(elementType, SignatureKind.Call);
+            if (callSignatures && callSignatures.length > 0) {
+                let callSignature: Signature;
+                callSignature = resolveCall(openingLikeElement, callSignatures, candidatesOutArray);
+                return callSignature;
+            }

```

Comment:
Why do we need to fill up `candidatesArray`? Is it for the language service? Shouldn't normal `resolveCall` fill up `candidatesArray` for a union?